### ISSUE TITLE
Passing config options for image build

### ISSUE THREAD

**Issue by rvs** [linuxkit/linuxkit#2817]
**Description**

Currently the kernel build section has a few hardcoded options for things like filer container, etc. It would be nice to have a configurable options that one can pass. The same mechanism could be used to potentially control things like GRUB, etc.

**Comment by errordeveloper** [comment]
@ijc and I just discussed this in the context of Kubernetes packages, where we have `kubernetes_version` and `cni_version` already, and it would be convenient to make it easier to pass other options, e.g. repo URL for developers to test a version in a fork...

So a natural thing to do seems to map it to `docker build --build-args`, aka `ARG` keyword in Dockerfile, and set defaults in `build.yaml`. The main concern is how to allow a developer to easily build packages with different parameters without breaking the mapping between git hash in the image tag and the content of the image.

One option I can think would involve:

- adding `args` field to `build.yml` to set default values
- adding `-args` flag to `linuxkit pkg build` to override the defaults set in `build.yml`

Let's consider how this could work nicely.

Given we have `pkg/kubelet/build.yml` that looks like this:

```yaml
org: linuxkit
image: kubelet
network: true
arches:
  - amd64
config:
  build_args:
    kubernetes_version: v1.9.0
    kubernetes_repo_url: git@github.com:kubernetes/kubernetes
```

When `linuxkit pkg build ./pkg/kubelet` runs, it produces `linuxkit/kubelet:<git_hash>` as it does at present.

However, if I were to run the following command instead:
```
linuxkit pkg build  -args \
   kubernetes_version=myfeature,kubernetes_repo_url=git@github.com:errordeveloper/kubernetes \
   ./pkg/kubelet
```
It would produce `linuxkit/kubelet:<git_hash>-<args_hash>`, where `args_hash` is the hash of `-args` and is also stored as an image label.

If we need a more formal design doc, I'm happy to produce such a thing, but let's discuss it first :)

**Comment by justincormack** [comment]
Yes, i discussed adding build args to build.yml with Roman and it seemed to
be a good solution, and doesnt break tree hash.

**Comment by rvs** [comment]
+1 @errordeveloper and @justincormack

**Comment by errordeveloper** [comment]
Should we have a more formal design doc for this?

**Comment by rvs** [comment]
I can certainly help -- but someone else will have to take a stab at it first.

**Comment by w9n** [comment]
If you do changes in `build.yml` the git subtree hash of the pkg dir changes so i dont think its necessary to introduce an args_hash.

**Comment by errordeveloper** [comment]
> If you do changes in `build.yml` the git subtree hash of the pkg dir changes.

@w9n the case that I am advocating here is about setting parameters on command line and not having to edit `build.yml`.

**Comment by errordeveloper** [comment]
@rvs I'm happy to take a stab at it.

**Comment by justincormack** [comment]
@errordeveloper I think it makes sense to support `build.yml` first as its a simple case

**Comment by errordeveloper** [comment]
@justincormack ah, sure, that is a natural place to start indeed.

**Comment by w9n** [comment]
the build_args (or config in general) could also be implemented into the tags field of `build.yml` in #2648, so that its reproducable from somewhere. This would require a sheme like @errordeveloper suggested as there could be different containers due to different configs on the same git tree.

**Comment by rvs** [comment]
Curious if there's been any progress on this since Jan. We're also now running into a related issue over here: https://github.com/zededa/zenbuild/issues/28

Although even if this issue gets fixed -- we still can't do the above transparently sine Docker doesn't allow variable substitution in FROM :-(

**Comment by justincormack** [comment]
You can do ARG substitution in FROM since sometime last year.

On 23 Feb 2018 21:09, "Roman V Shaposhnik" <notifications@github.com> wrote:

Curious if there's been any progress on this since Jan. We're also now
running into a related issue over here: zededa/zenbuild#28
<https://github.com/zededa/zenbuild/issues/28>

Although even if this issue gets fixed -- we still can't do the above
transparently sine Docker doesn't allow variable substitution in FROM :-(

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/linuxkit/linuxkit/issues/2817#issuecomment-368139249>,
or mute the thread

**Comment by deitch** [comment]
> If you do changes in build.yml the git subtree hash of the pkg dir changes.

Which implies that we shouldn't need something like `linuxkit/kubelet:<git_hash>-<args_hash>`, since the tree hash changes.

The challenge is passing args via CLI. For that, I don't think anything is reproducible. With `tree hash`, I always can find the specific tree hash in the git history. If I do `args hash`, I cannot get to those args again. It just isn't reproducible. That leaves a few options:

* Use just `tree hash`: really bad idea, since it implies something that it isn't
* Do not support `-args`, requiring people to change the `build.yml`
* Indicate that it is dirty, which can be handled equally with `-dirty` as `-<args_hash>`

I think `-dirty` is the way we have been indicating that "this is not something you can reproduce unless you really kept track of it, don't look to git to find it." On the other hand, do we not store the `linuxkit.yml` in the image somewhere? Could we store the args and `build.yml`, and then it would be sufficient to indicate `-dirty` or something similar?

**Comment by justincormack** [comment]
Storing the config in the image is supported. It was at one point missing some of the info but I think thats fixed now.

Personally I would always build something for production from git, even if the `build.yml` files were generated I would still be inclined to commit them to a repo for reference. A hash of some arguments is no better than `dirty` as finding the arguments after is tedious trying to find what hashed to that value.

**Comment by deitch** [comment]
> Storing the config in the image is supported

So `build.yml` for package OCI images and `linuxkit.yml` for lkt images?

> A hash of some arguments is no better than dirty as finding the arguments after is tedious trying to find what hashed to that value

That makes sense. I think it isn't so much of what is possible, as the process and behaviour it creates. Calling something `-dirty` just makes people think, "ah, this is dirty, good enough for now, but I don't want that word on my image. How do I get rid of it? Oh, right, commit to git."

**Comment by rvs** [comment]
I like this approach very much: -args always resulting in -dirty makes perfect sense to me and would allow for the use case that originally started this issue.
